Rewrite security section (#322)
akeneo/pim-api-docs: content/rest-api/authentication.md

1. Input a label for your connection, `ERP` for example.
1. Select a type flow. [_Don't know what it is? More info here._](https://help.akeneo.com/pim/articles/manage-your-connections.html#choose-your-flow-type)
1. Click on the `Save` button.

Congrats! You officially created your very first connection to the PIM. :tada:

You can now find the client ID and secret in the `Credentials` section of the connection you've just created.

[Not comfortable with the **Connection** notion? We explain everything right here!](https://help.akeneo.com/pim/articles/what-is-a-connection.html)

::: info
In case your secret leaked, you can easily revoke it and generate a new one for your connection by clicking on the dedicated icon, on the right side of the secret.
:::

### With a v2.x and 3.x PIM

In 2.x or 3.X, you will need to create what we called an `API connection`. 

1. Log into your favorite PIM.
1. Navigate to the `System/API connections` menu. 
1. Click on the `Create` button. 
1. Input a label for your connection, `ERP connection` for example.
1. Click on the `Save` button.

The PIM automatically generates a client ID and secret, that you will find in the API connections grid.

::: info

> In case your secret leaked, you can revoke an API connection by clicking on the `Revoke` button, in the `API connections` entry you will find in the `System` menu.  

Author: ```suggestion
In case your secret leaked, you can revoke an API connection by clicking on the `Revoke` button. It is located in the `API connections` entry you will find under the `System` menu.  
```

| 4_5e6kfzmath8gowk0s000kkgc0o44cwgwsockwk0ccss4sw0w | 2nwha9mzk2w4so0cgokwocswoc48s0sg44wgg40kkokgg4w0go | Magento connection |
+----------------------------------------------------+----------------------------------------------------+-------------------+
```

## API user creation

You now have a client ID and a secret.  But the request to get this token needs a username and a password.

### Starting from the v4 of the PIM

In v4 PIM, this username and password are automatically generated whenever you create a connection in the UI. Navigate to `System/Connections`, click on your connection and find your username and password in the `Credentials` section.

You may need to regenerate a password if you haven't saved it anywhere. Indeed, for security reasons, the PIM can only show it to you once, after the connection is created. There is a `Regenerate` button close to the hidden password if needed. :wink:

### With a v1.7, 2.x and 3.x PIM

In previous versions of the PIM, you need to navigate to the `System/Users` menu in your PIM UI and create a new user. We strongly advise you to create specific API users. Do not re-use UI users. 🙂

## Token generation

> You now have a client ID, a secret, a usename and a password. The last information needed to use the API is the token.

Author: ```suggestion
You now have a client ID, a secret, a usename and a password. The last information needed to use the API is the token.
```
```suggestion
You now have a client ID, a secret, a username and a password. The last information needed to use the API is the token.
```

> First, you will have to encode in base64 the secret id and the secret given by the administrator with a `:` in between.

Author: ```suggestion
First, you will have to encode the client ID and the secret, in base64, with a `:` in between.
```

:::

#### Example

**Request**
```bash
curl -X POST http://my-favorite-pim.com/api/oauth/v1/token \
    -H "Content-Type: application/json" \
    -H "Authorization: Basic NGdtNHJub2l6cDhnc2tna2swODBzc29vODAwNDBnNDRrc293d2d3ODQ0azQ0c2MwMHM6NWR5dm8xejZ5MzRzbzRvZ2tna3N3ODhvb2tvb3dzMDBjZ29jNDg4a2NzOHdrNGM0MHM=" \
    -d '{
        "grant_type": "password",
        "username": "myERPuser",
        "password": "64bngr78"
    }'
```

**Response**
```http
HTTP/1.1 200 OK

{
    "access_token": "NzFiYTM4ZTEwMjcwZTcyZWIzZTA0NmY3NjE3MTIyMjM1Y2NlMmNlNWEyMTAzY2UzYmY0YWIxYmUzNTkyMDcyNQ",
    "expires_in": 3600,
    "token_type": "bearer",
    "scope": null,
    "refresh_token": "MDk2ZmIwODBkYmE3YjNjZWQ4ZTk2NTk2N2JmNjkyZDQ4NzA3YzhiZDQzMjJjODI5MmQ4ZmYxZjlkZmU1ZDNkMQ"
}
```

Detail of the information given in the response:
- `access_token` must be included in every request to the API for the client application to be authorized (see the example below).
- `expires_in` is the token lifespan (in seconds). By default, it lasts 1 hour.
- `refresh_token` is a special token used only to refresh your `access_token` after it expired.

## Access a resource

> Use the token that you got before in every request to the API.

Author: ```suggestion
Use the token that you just got for every request to the API.
```

After a token has expired, the client application can ask for a new token by calling:

```bash
curl -X POST http://your-host/api/oauth/v1/token\
    -H "Content-Type: application/json" \
    -H "Authorization: Basic YOUR_BASE_64_CLIENT_ID_AND_SECRET" \
    -d "grant_type"=refresh_token \
    -d "refresh_token"=REFRESH_TOKEN
```

#### Example

**Request**
```bash
curl -X POST http://your-host/api/oauth/v1/token \
    -H "Content-Type: application/json" \
    -H "Authorization: Basic M2UyaXFpbHEyeWd3azBjY2dvZ2tjd2NvOG9vc2Nra2trNGdrb2MwazRzOHMwNDR3c3M6NDRlY3Rlbm11ZHVzOGc4OHc0d2t3czg0MDQ0Y2t3MGs0dzRrZzBzb2tvc3M4NG9rbzg=" \
    -d "grant_type"=refresh_token \
    -d "refresh_token"=MDk2ZmIwODBkYmE3YjNjZWQ4ZTk2NTk2N2JmNjkyZDQ4NzA3YzhiZDQzMjJjODI5MmQ4ZmYxZjlkZmU1ZDNkMQ 
```

**Response**
```http
HTTP/1.1 200 OK

{
  "access_token": "MjkxYjM0ZGU0OGQzN2MzNmRmOGFkMzgxODdhMDZiNmE1MTI2ZDJmYmM5MTBhZWFiMGY0ZjE3YzJhOTJkOGRlZA",
  "expires_in": 3600,
  "token_type": "bearer",
  "scope": null,
  "refresh_token": "MTU2OWI5NGYwZTU2MThmMjE3MzRmYmIyMjlmYzU5Zjc3ZDhlM2E5YzQyMmIyMDVhOGY2OWQ3MTM0MTMzNmZlYw"
}
```

> The response is the same as when asking for the initial token except that it contains a new `access_token` and `refresh_token`. You can then use the new `access_token` in every request you make to the API.

Author: ```suggestion
The response is the same as when asking for the initial token, except that it contains a new `access_token` and a new `refresh_token`. You can then use this `access_token` for every API request you do within an hour.
```

| 4gm4rnoizp8gskgkk080ssoo80040g44ksowwgw844k44sc00s | 5dyvo1z6y34so4ogkgksw88ookoows00cgoc488kcs8wk4c40s | ERP connection |
| 4_5e6kfzmath8gowk0s000kkgc0o44cwgwsockwk0ccss4sw0w | 2nwha9mzk2w4so0cgokwocswoc48s0sg44wgg40kkokgg4w0go | Magento connection |
+----------------------------------------------------+----------------------------------------------------+-------------------+
```

## API user creation

You now have a client ID and a secret.  But the request to get this token needs a username and a password.

### Starting from the v4 of the PIM

In PIM v4, this username and password are automatically generated whenever you create a connection in the UI. Navigate to `System/Connections`, click on your connection and find your username and password in the `Credentials` section.

You may need to regenerate a password if you didn't save it somewhere. Indeed, the PIM can only show it to you once, at the connection creation for security reasons. There is a `Regenerate` button close to the hidden password, if needed. :wink:

### With a v1.7, 2.x and 3.x PIM

> In previous versions of the PIM, you need to navigate to the `System/Users` menu in your PIM UI and create a new user. We strongly advise you to create specific API user. Do not re-use UI users. 🙂

Reviewer: ```suggestion
In previous versions of the PIM, you need to navigate to the `System/Users` menu in your PIM UI and create a new user. We strongly advise you to create specific API users. Do not re-use UI users. 🙂
```

## Token generation

> You now have a client ID, a secret, a usename and a password. The last information needed to use the API is the token.

Reviewer: username ?
